File: backend/src/dataclasses/playback_request.py
from pydantic import BaseModel
from typing import List, Literal, Optional, Union


class Offset(BaseModel):
    type: Literal["position", "uri", "album_id"]


class PositionOffset(Offset):
    type: Literal["position"] = "position"
    position: int


class UriOffset(Offset):
    type: Literal["uri"] = "uri"
    uri: str


class AlbumIdOffset(Offset):
    type: Literal["album_id"] = "album_id"
    album_id: str


class StartPlaybackRequest(BaseModel):
    context_uri: Optional[str] = None
    uris: Optional[List[str]] = None
    offset: Optional[Union[PositionOffset, UriOffset, AlbumIdOffset]] = None
    position_ms: Optional[int] = None


class ResumePlaybackRequest(BaseModel):
    id: str
    context_uri: Optional[str] = None

File: playlist-manager-next/app/utils/interfaces/PlaybackRequest.ts
export type OffsetType = "position" | "uri" | "album_id";

export interface Offset {
  type: OffsetType;
}

export interface PositionOffset extends Offset {
  type: "position";
  position: number;
}

export interface UriOffset extends Offset {
  type: "uri";
  uri: string;
}

export interface AlbumIdOffset extends Offset {
  type: "album_id";
  album_id: string;
}

export type PlaybackOffset = PositionOffset | UriOffset | AlbumIdOffset;

export interface StartPlaybackRequest {
  context_uri?: string;
  uris?: string[];
  offset?: PlaybackOffset;
  position_ms?: number;
}

export interface ResumePlaybackRequest {
  id: string;
  context_uri?: string;
}
